examples/rvc-app/rvc-common/src/rvc-device.cpp

```diff
@@ -0,0 +1,107 @@
+#include "rvc-device.h"
+
+using namespace chip::app::Clusters;
+
+void RvcDevice::Init()
+{
+    mRunModeInstance.Init();
+    mCleanModeInstance.Init();
+    mOperationalStateInstance.Init();
+}
+
+void RvcDevice::HandleRvcRunChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t currentMode = mRunModeInstance.GetCurrentMode();
+
+    switch (NewMode)
+    {
+    case RvcRunMode::ModeMapping: {
+        // change to mapping only allowed from the Idle state.
+        if (currentMode != RvcRunMode::ModeIdle)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString("Change to the mapping mode is only allowed from idle"));
+            return;
+        }
+
+        mRunModeInstance.UpdateCurrentMode(NewMode);
+        // Set operational state to running
+        ChipError err =
+            mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+        if (err != CHIP_NO_ERROR)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString(err.AsString()));
+            return;
+        }
+        break;
+    }
+    case RvcRunMode::ModeCleaning: {
+        mRunModeInstance.UpdateCurrentMode(NewMode);
+        // Set operational state to running
+        ChipError err =
+            mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+        if (err != CHIP_NO_ERROR)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString(err.AsString()));
+            return;
+        }
+    }
+    break;
+    case RvcRunMode::ModeIdle: {
+        mRunModeInstance.UpdateCurrentMode(NewMode);
+        // Set operational state to running
+        ChipError err =
```

Comment:
Perhaps add a TODO comment about generating a completion event for the Running/Paused to Idle transitions.

Reply:
This will need to involve either timers or out-of-bound messages. I was thinking of building another app that could go in all the states via out-of-bound messages and keep this simple and with few dependencies.